src/resources/ble_esp32.cc

```diff
@@ -18,721 +18,1682 @@
+      break;
+  }
+  String* str = process->allocate_string(buffer, static_cast<int>(strlen(buffer)));
+  if (!str) ALLOCATION_FAILED;
+  return Primitive::mark_as_error(str);
+}
 
-  ble_pool.put(_id);
+Object* nimle_stack_error(Process* process, int error_code) {
+  return nimble_error_code_to_string(process, error_code, false);
 }
 
 
-static ble_uuid_any_t uuid_from_blob(Blob blob) {
-  ble_uuid_any_t uuid = { 0 };
-  if (memcmp(kBluetoothBaseUUID+4, blob.address()+4, 12) == 0) {
-    // Check if it's 16 or 32 bytes.
-    if (memcmp(kBluetoothBaseUUID, blob.address(), 2) == 0) {
+static ble_uuid_any_t uuid_from_blob(Blob& blob) {
+  ble_uuid_any_t uuid = {};
+  switch (blob.length()) {
+    case 2: {
       uuid.u.type = BLE_UUID_TYPE_16;
-      uint16 value = *reinterpret_cast<const uint16*>(blob.address() + 2);
+      uint16 value = *reinterpret_cast<const uint16*>(blob.address());
       uuid.u16.value = __builtin_bswap16(value);
-    } else {
+      break;
+    }
+    case 4: {
       uuid.u.type = BLE_UUID_TYPE_32;
       uint32 value = *reinterpret_cast<const uint32*>(blob.address());
       uuid.u32.value = __builtin_bswap32(value);
+      break;
     }
-  } else {
-    uuid.u.type = BLE_UUID_TYPE_128;
-    memcpy_reverse(uuid.u128.value, blob.address(), 16);
+    default:
+      uuid.u.type = BLE_UUID_TYPE_128;
+      memcpy_reverse(uuid.u128.value, blob.address(), 16);
   }
   return uuid;
 }
 
-uint32_t BLEServerConfigGroup::on_event(Resource* resource, word data, uint32_t state) {
-  switch (data) {
-    case BLE_GATT_ACCESS_OP_READ_CHR:
-      state |= kBLECharAccessed;
+static ByteArray* byte_array_from_uuid(Process* process, ble_uuid_any_t uuid, Error*& err) {
+  err = null;
+
+  ByteArray* byte_array = process->object_heap()->allocate_internal_byte_array(uuid.u.type/8);
+  if (!byte_array) {
+    err = Error::from(process->program()->allocation_failed());
+    return null;
+  }
+  ByteArray::Bytes bytes(byte_array);
+
+  switch (uuid.u.type) {
+    case BLE_UUID_TYPE_16:
+      *reinterpret_cast<uint16*>(bytes.address()) = __builtin_bswap16(uuid.u16.value);
       break;
-    case BLE_GATT_ACCESS_OP_WRITE_CHR:
-      state |= kBLECharReceived;
+    case BLE_UUID_TYPE_32:
+      *reinterpret_cast<uint32*>(bytes.address()) = __builtin_bswap16(uuid.u32.value);
       break;
     default:
-      break;
+      memcpy_reverse(bytes.address(), uuid.u128.value, sizeof(uuid.u128.value));
   }
-  return state;
+
+  return byte_array;
 }
 
-void BLEServerConfigGroup::set_subscription_status(uint16 attr_handle, uint16 conn_handle, bool indicate, bool notify) {
-  for (auto service: _services) {
-    for (auto characteristic: service->characteristics()) {
-      if (characteristic->nimble_value_handle() == attr_handle) {
-        characteristic->set_subscription_status(indicate, notify, conn_handle);
-        return;
-      }
-    }
+static bool uuid_equals(ble_uuid_any_t& uuid, ble_uuid_any_t& other) {
```

Comment:
Maybe.. Not even quite sure that "normal" implementors even consider that the 16 and 32 bit versions are abbreviated full uuids. So it might raise confusion.